oak/client/sgx_application_client.cc

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2019 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "oak/client/sgx_application_client.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "asylo/grpc/auth/enclave_channel_credentials.h"
+#include "asylo/grpc/auth/enclave_credentials_options.h"
+#include "asylo/grpc/auth/null_credentials_options.h"
+// TODO: Uncomment when Asylo will release remote attestation.
+// https://github.com/project-oak/oak/issues/119
+//#include "asylo/grpc/auth/sgx_remote_credentials_options.h"
+#include "asylo/identity/descriptions.h"
+#include "asylo/identity/enclave_assertion_authority_configs.h"
+#include "asylo/identity/init.h"
+#include "asylo/identity/sgx/sgx_identity_util.h"
+#include "asylo/identity/util/sha256_hash.pb.h"
+#include "asylo/util/logging.h"
+#include "asylo/util/status.h"
+#include "asylo/util/statusor.h"
+#include "oak/client/authorization_bearer_token_metadata.h"
+#include "oak/client/policy_metadata.h"
+#include "oak/common/nonce_generator.h"
+#include "oak/common/policy.h"
+
+using asylo::Status;
+using asylo::StatusOr;
+
+namespace oak {
+
+namespace {
+
+constexpr size_t kPerChannelNonceSizeBytes = 32;
+// Debug MRSIGNER value derived from the SGX test key (will change after changing SVN and PRODID).
+// https://github.com/google/asylo/blob/088ea3490dd4579655bd5b65b0e31fe18de7f6dd/asylo/distrib/sgx_x86_64/linux_sgx_2_6.patch#L5481
+const char* kDebugMrSigner = "83d719e77deaca1470f6baf62a4d774303c899db69020f9c70ee1dfc08c7ce9e";
+// Debug CPUSVN value.
+// This value should be provided by EREPORT and it reflects the microcode update version.
+const char* kDebugCpuSvn = "0000000000000000";
+
+// Converts a hexadecimal string `hash_string` to the string of bytes.
+// This function is a wrapper for an absl::HexStringToBytes with a validity check.
+// TODO: Use same function from Asylo, when it becomes public.
+StatusOr<std::string> HexStringToBytes(const std::string& hash_string) {
+  for (auto ch : hash_string) {
+    if (std::isxdigit(ch) == 0) {
+      return Status(asylo::error::GoogleError::INVALID_ARGUMENT,
+                    "String contains non-hexademical charachters: " + hash_string);
+    }
+  }
+  return absl::HexStringToBytes(hash_string);
+}
+
+// Parses a hexadecimal hash string from `hash_string` into `Sha256HashProto` in `hash`.
+// `hash_bytes` should be a hexadecimal string that represents a Sha256 hash (64 bytes long).
```

Comment:
64 or 32 bytes? The wording is ambiguous, whether it's referring to the string or the hash :)

Reply:
Added clarification